### DISCUSSION TITLE
Getting started with CompositorEffects and Compute shaders

### DISCUSSION THREAD

**Discussion by pink-arcana** [pink-arcana/godot-distance-field-outlines#1] [General]
Distance Field Outlines was a learning project for me. I've been programming for years, but not professionally or at a low-level. Before starting this, I had some experience with simple Godot shaders, and had set up some multi-node post-processing effects. But I'd never touched a Compute shader or Godot's lower-level graphics APIs. I say this to show that you *can* get started creating CompositorEffects without having a background in graphics programming. However, you do have to be willing to do a lot of your own homework, and be okay with not being able to find quick solutions by googling.

I took lots of notes as I was making this project. I initially wrote them in comments inside my scripts, so now that I'm cleaning the project up to post the source code, I'm adding those notes into this discussion thread below. (Edit: I'm also adding and correcting info and links as I learn more.)

But, like I said, I'm not an expert! So take any advice I give with a grain of salt. I welcome feedback or corrections. And I'd also love to get your own tips and links, both for myself and anyone else visiting.

All my benchmarks here are from an Nvidia RTX 3080 laptop GPU. If you test any of the same things I did on different hardware and get different results, please post your findings! I opened a separate discussion for [Performance & Optimization](https://github.com/pink-arcana/godot-distance-field-outlines/discussions/5) as well.

### Getting started
- [Godot's NEW feature is AWESOME](https://www.youtube.com/watch?v=rVcd8sLPsXU) by Octodemy - YouTube tutorial/introduction to CompositorEffects (thanks to @jasonswearingen for the link)
- Godot docs : [The Compositor](https://docs.godotengine.org/en/latest/tutorials/rendering/compositor.html) - official tutorial and starter project.
- Bastiaan Olij's dev streams: [49](https://www.youtube.com/live/insa9VY0ruA) and [50](https://www.youtube.com/watch?v=KLmahauWeV4) (follow the links to the demo projects in the descriptions) - This is how I got started with CompositorEffects. He created the CompositorEffect PR, and it's super helpful to hear him describe, line-by-line, what the code in the demo projects is doing and why.
- [The original CompositorEffect PR](https://github.com/godotengine/godot/pull/80214) - the discussion was helpful for understanding what it can/can't do, and contains some useful tips.
- [RenderDoc](https://renderdoc.org/) - absolute must-have for debugging. You can learn a lot about how the rendering works in Godot just by snooping around inside a captured frame.
	- To get labels to work, use `--verbose`
- [BaseCompositorEffect](https://github.com/pink-arcana/godot-distance-field-outlines/blob/main/project/addons/pa_compositor_effect/base_compositor_effect.gd) (see also: [BaseCompositorEffect script template](https://github.com/pink-arcana/godot-distance-field-outlines/blob/main/project/script_templates/CompositorEffect/base_compositor_effect.gd) and [SceneData ](https://github.com/pink-arcana/godot-distance-field-outlines/blob/main/project/df_outline_ce/shaders/includes/scene_data.glsl) and [SceneDataHelper](https://github.com/pink-arcana/godot-distance-field-outlines/blob/main/project/df_outline_ce/shaders/includes/scene_data_helpers.glsl) shader includes) - I refactored my CompositorEffect for this project into a superclass that abstracts a lot of the repetitive code. This is specific to my workflow, but may be helpful to look at. It includes functions for binding the build-in SceneData UBO and the built-in depth and normal/roughness textures.

#### Vulkan references
The GDScript naming is slightly different from the Vulkan API, and RenderingDevice also abstracts much of the Vulkan setup. So, when using Vulkan references, I've found it helpful to have the GDScript class references and C++ code open side-by-side to figure out how it corresponds.

<details><summary>Quick links</summary>
<p>

- RenderingDevice: [GDScript](https://docs.godotengine.org/en/latest/classes/class_renderingdevice.html) | [Source](https://github.com/godotengine/godot/blob/master/servers/rendering/rendering_device.cpp)
- RenderSceneBuffersRD: [GDScript](https://docs.godotengine.org/en/latest/classes/class_renderscenebuffersrd.html) | [Source](https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.cpp)
</p>
</details> 

- [Vulkan documentation](https://docs.vulkan.org/spec/latest/index.html) - much more approachable than the specs, and you can read [the guide section](https://docs.vulkan.org/guide/latest/index.html) like a manual. Oddly, doesn't seem to show up google search results, even though it has the official answers for many common questions. There is also [a re-organized version](https://vkdoc.net/chapters/introduction) here, with better search.
- [vkguide.dev](https://vkguide.dev) - excellent introduction to every Vulkan topic you'll need to understand for CompositorEffects. It contains specific guidance on best practices (I linked to a couple specific ones in the sections below.)
- [Vulkan Tutorial](https://vulkan-tutorial.com) - mostly covers the same topics as vkguide.dev, but reading them explained in a different way can make it click.

**Comment by pink-arcana** [comment]
### Storage Options
These are the non-texture uniform options I found useful for this project, and my experiences with them. Also see the [Textures post below](https://github.com/pink-arcana/godot-distance-field-outlines/discussions/1#discussioncomment-10016608).

For a full list of options and GLSL examples, see docs.vukan.org - [Mapping Data to Shaders](https://docs.vulkan.org/guide/latest/mapping_data_to_shaders.html).

- **Push constants** - push a small amount of data to a shader at the time of compute list creation. GPUs should be transporting and storing them in an optimized way for fast access.
	- Individual elements must be aligned to 4 bytes, and the entire push constant must have a size that's a multiple of 16 bytes. They have a maximum size of 128 bytes in Godot.
	- Used in the Godot docs example projects.

- **Uniform Buffer Objects (UBOs)** - move a larger amount of data for persistent storage on the GPU. Then give your shaders access to it via bindings in their layouts.
	- Aligned by 16 bytes. Can be much larger than push constants on desktop GPUs, but UBOs larger than 128 bytes may not be reliable across mobile hardware. (Arm specifically states they limit to 128 bytes.)
		- If you assign the members inside of the UBO layout, they must be aligned per the layout to 16 bytes (i.e. they must all be at least vec4's). However, if you assign the whole object to a separate struct, you can align elements by 4 bytes -- with some exceptions. This packs the data more tightly, and has both pros and cons for usability. Both options had equivalent performance in my testing.
			- The exceptions are:
				- Larger elements must still be aligned to 16 bytes. So I've found it easier to put all the vec4s first, then the floats.
				- While you can assign floats directly in this struct, I got bad data when I tried to assign bools or ints. However, Godot's SceneData struct directly assigns bools and uints (presumably without a problem?). So this warrants further investigation.
		- See this thread on Reddit with helpful replies from a Godot maintainer: [Uniforms vs buffers in compute shaders, and how do I bind them?](https://www.reddit.com/r/godot/comments/1510ge1/uniforms_vs_buffers_in_compute_shaders_and_how_do/)
	- It seems that only static UBOs are available in Godot 4.3, so updating a value requires re-creating the UBO. In this project, that doesn't seem to be a problem. It updates the UBO whenever the settings change, but otherwise just hangs out in GPU memory.
	- You can also connect your shaders to Godot's internal SceneData UBO. It is already in memory, so it gives you access to handy data (like viewport size, transforms, time) essentially for free.
		- Instructions for binding it are in [the Github comments here](https://github.com/godotengine/godot/pull/80214#issuecomment-1953258434)).
		- The [struct itself is here](https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/scene_data_inc.glsl).
	- In testing for this project, binding to pre-existing UBOs was faster than using a push constant. The push constant had only been pushing over the specific data the shader needed, while each UBO now provides each shader with extra data it doesn't use. (For SceneData, this is a struct of almost 100 lines.) However, these unused assignments do not seem to matter. So, from what I can tell, binding existing UBOs (like SceneData) should probably be preferred to sending the same data over in pieces with push constants.
	- You can debug your UBOs in RenderDoc: Under the Compute Pass, select the appropriate vkCmdDispatch. Then open the Pipeline tab, and click the Compute Shader button. In the Uniform Buffers list, double-click the item to see its values.

- **Specialization constants** - add small, individual constants directly to a pipeline when you create it. The shader is then compiled on the GPU with these constants hard-coded in. The constants are very limited in number and size, but using them can be very performative. This is a reason some games have massive numbers of pipelines (see [this screenshot from Breath of the Wild](https://www.reddit.com/r/cemu/comments/14o230h/loading_vulkan_pipelines_in_breath_of_the_wild/).)
	- There seem to be two main use cases:
		- **Hard-coding values that differ by shader pass.** - In my testing, creating individual pipelines for each JFA pass with the offset value as a specialization constant was slightly faster than using push constants (-80μs over 11 passes).
		- **Replacing constant uniforms used for branching.**
			- Modern desktop GPUs have a concept of constant uniforms, so, if the uniform has the same value for the entire workgroup, it is treated as a constant. Specialization constants would not have any performance benefit in this case. And, in my testing on my Nvidia RTX GPU with a highly-branching shader, the performance was the same either way.
			- However, on GPUs that do not have constant uniforms (for example, Arm states their chips do not), the difference may be significant. This leads me to think that, if both options are otherwise equivalent, specialization constants should be preferred.
	- Note that while specialization constants are compiled into the shader by the GPU, they are *not* considered compile-time constants (at least in Godot's GLSL compiler), so cannot be used to set offsets for texture + offset commands. They also can't be used to set array lengths as of 4.2/4.3 ([issue](https://github.com/godotengine/godot/issues/94881))

**A note on layout requirements**
Push constants and UBOs are specific about their alignment and often require extra padding. I personally found it easier and less error-prone to abstract out the alignment to separate functions. You can compile and run your shaders without error even if you do not assign the extra padding that's been added. This makes it faster to quickly add/remove uniforms for testing. However, this seems like it could be creating a buffer overflow, so, once I was done setting up, I still added in extra floats to capture the unused data.

**Comment by pink-arcana** [comment]
### Uniform sets and bindings

- [vkguide.dev : Descriptor sets](https://vkguide.dev/docs/chapter-4/descriptors/) - excellent guide to what sets are, and how to optimally set them up, including when to increment set index versus binding. They point out that some hardware only supports a maximum of 4 layout sets. And they suggest organizing sets in an order of most-global to most-local.
- [zeux.io - Writing an efficient Vulkan renderer](https://zeux.io/2020/02/27/writing-an-efficient-vulkan-renderer/) - goes into much more detail about different options for organizing sets.

**Comment by pink-arcana** [comment]
### Compute shaders

- [vkguide.dev : Compute shaders](https://vkguide.dev/docs/gpudriven/compute_shaders/) - introduction to how compute shaders work, and how to choose a workgroup size.
	- The Godot examples use a workgroup size of 8x8.
	- Per vkguide.dev, "It’s very common to have things like 16x16 workgroup size when doing postprocess filters or other similar rendering shaders."
	- I chose a workgroup size of 16x16 for this CompositorEffect based on my results here:

		    Total frame times for different sizes @ 1080p/1024px:
			    8x8 -> 1.43ms
		 	    16x16 -> 1.15ms
			    32x32 -> 1.26ms
		 	    64x64 -> error: command_queue_execute_and_present() FAILED

- gl_GlobalInvocationID gives you the texel's coordinates in texture space. It is located at the edge of the texel, not the center. To convert to a normalized position to use in texture() functions, you need to add 0.5. See [this stack overflow answer](https://stackoverflow.com/questions/40574677/how-to-normalize-image-coordinates-for-texture-space-in-opengl) for reference.

**Comment by pink-arcana** [comment]
### Textures
- NekotoArts's [Everything About Textures in Compute Shaders!](https://nekotoarts.github.io/blog/Compute-Shader-Textures) - a Godot-specific guide to creating your own textures and moving back and forth from CPU to GPU.

#### Choosing a format
- [Vulkan: optimal image tiling format support on all platforms](https://vulkan.gpuinfo.org/listoptimaltilingformats.php) - Comprehensive list of formats sorted by % support for each platform.

#### Choosing a uniform
Use **image** uniforms with imageStore/imageLoad to read and/or write a single texel; use **sampler** uniforms with the GLSL `texture` functions to read.
- You can find many opinions about which is better/faster.
	- Arm's guide says, [`Use texelFetch() or texture() instead of the slower imageLoad() wherever possible.`](https://developer.arm.com/documentation/101897/0302/Buffers-and-textures/Texture-sampling-performance)
	- On my Nvidia RTX GPU, head-to-head benchmarks of imageLoad vs texelFetch were identical.

#### Built-in textures
- Normal roughness texture
	- After sampling, apply [`normal_roughness_compatibility()` function](https://github.com/godotengine/godot/blob/685413a9489fb31f897770d6619a45b5a9c6e6a8/servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl#L334) (From [Use best fit normals for storing screen space normals](https://github.com/godotengine/godot/pull/86316) PR)
		- Because this was so hard to find, I [opened an issue](https://github.com/godotengine/godot-docs/issues/9591) for this on the Godot docs repo
- Depth does not have necessary usage bits to copy from or to it, so it cannot be updated from a CompositorEffect ([source](https://github.com/godotengine/godot/blob/0e9caa2d9cb20737f8dcf08b75fcf2a78d980569/servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.cpp#L207))


#### Sharing textures between different CompositorEffects
- For two CompositorEffects in the same viewport, `render_scene_buffers.has_texture()` is **true** for textures created in the other CompositorEffect.
- For CompositorEffects in different viewports, `render_scene_buffers.has_texture()` is **false** for textures created in the other CompositorEffect.
	- However, the texture image still exists in memory. You can access it by passing its RID to the other CompositorEffect (tested with setting a variable on an autoload).

**Comment by pink-arcana** [comment]
### CompositorEffects

#### Troubleshooting
- If the CompositorEffect is in a scene without an Environment (on a Camera3D or WorldEnvironment), `_render_callback()` will be called early, but still pass the expected `EffectCallbackType`. For me, this has resulted in errors accessing the `normal_roughness` buffer, since it doesn't exist yet. But this is a red herring. The true problem is the lack of an Environment in the scene.
- A CompositorEffect's `@export` variables cannot be keyframed in the AnimationPlayer from inside the Compositor's `Array[CompositorEffect]`. A workaround is to put the same saved CompositorEffect resource (or a separate settings resource) directly on a node in the scene tree. [See the instructions here](https://github.com/pink-arcana/godot-distance-field-outlines/blob/main/project/df_outline_ce/README.md#animating).
- The `normal_roughness` texture will appear grainy or degraded and cause artifacts when used for rendering. [It requires a conversion function](https://github.com/godotengine/godot-docs/issues/9591).
- If you are reading and writing to the same image from a single Compute shader, you can only read from the same texel you are writing to. Sampling neighbors will result in flickering and other artifacts due to the parallel processing. The built-in textures (like the screen color image) don't seem to have the TEXTURE_USAGE_CAN_COPY_FROM_BIT set. So you would need to create a custom copy to a new texture in a separate shader pass (see an example of how to do this [in this tutorial](https://youtu.be/rVcd8sLPsXU?si=37jQ_TcjS8tAlEpr&t=318)).

**Comment by jasonswearingen** [comment]
Here is a pretty nice video intro to compositor effects?   https://youtu.be/rVcd8sLPsXU?si=fttfutAfDIqbZDPf

**Reply by pink-arcana** [reply]
Very nice! I added it to the top of the list since it looks like a much easier place to start than anything else on there.

**Comment by Mopzilla** [comment]
Just wanted to say this information is extremely helpful. I'm running into limitations with standard shaders and was looking to move into using the compositor and this has helped a bunch. Great post 👍

**Reply by jasonswearingen** [reply]
Agree!  I'm super excited to have this available.   I'm finally getting to my visuals (procedural spaceship hull) so this bookmark is warm and ready :)

**Comment by thygrrr** [comment]
How do I copy JUST a (viewport)-texture into my framebuffer, basically initializing it with the output of another camera?

Can't find even the simplest example how to draw an image in Compositor.

**Reply by KerekesDavid** [reply]
Did you figure this out? I'm trying to bind a ViewportTexture to a shader in a compositor effect, but all I get is errors saying it's not a valid texture.

**Reply by thygrrr** [reply]
I did, in fact, not figure it out.

**Reply by obsurveyor** [reply]
> Did you figure this out? I'm trying to bind a ViewportTexture to a shader in a compositor effect, but all I get is errors saying it's not a valid texture.
@thygrrr 

In case you still haven't figured it out and for anyone else looking:
`var input_viewport: RID = RenderingServer.texture_get_rd_texture(viewport.get_texture())`

`viewport` is just a reference to the `SubViewport`, which is kind of a pain to get into the Compositor Effect script. I haven't figured out a great way to do it yet. Right now I have a script on the `SubViewport` that sets the variable on the Compositor Effect instance directly, which is real brittle.
 
Assuming the `SubViewport` is in the same instance as the `WorldEnvironment` with the Compositor Effect, you need to set a 'World Environment' and empty(or not) `Compositor` on the `SubViewport` itself or you'll just get a black texture.  Also may want to set the `Use HDR 2D` under 'Rendering'.

Then bind it to a uniform and use it like you would any other input texture in a compute shader. Don't forget the push constants for the size of the `SubViewport` if it isn't the same size as your main render texture.

**Reply by jovlem** [reply]
@obsurveyor
I'm trying to get a SubViewport working inside a compositor effect, but can't get it to work.
So I need to set a WorldEnvironment as a child to the SubViewport? Or should I set the SubViewport as I child of the WorldEnvironment ( or maybe it does not matter which one is parent or child )? 

And then I also need to add a Compositor effect to the SubViewport.
But do I only need to add an element to the array, so basicly just pressing the "Add Element" button. Or do I also need to create an extra script and attach that as well ( and also create a shader for it )?

And do you use an extra camera in the SubViewport ( that's what I have now ), or maybe a ColorRect?

**Comment by obsurveyor** [comment]
I've [collected](https://codeberg.org/obsurveyor/godot_compositor_effects_collected) a number of Compositor Effects and information about them, including a link to this discussion, into a repository for those who want to learn more about Compositor Effects with examples.  The info and effects are spread out across the Internet, so I thought consolidating them would be useful. It will be updated over time as I learn more and find or write more examples to include.

**Comment by robostep9829** [comment]
Can this be used in mobile renderer for _anything_? Since the color texture there doesn't have the storage flag, it can only be written by fragment shaders, not compute. Also it doesn't have the "copy to" flag neither in forward+ nor mobile renderer.
